subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilationInitializer.java

```diff
@@ -52,13 +56,20 @@ public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recomp
         PatternSet classesToDelete = patternSetFactory.create();
         PatternSet sourceToCompile = patternSetFactory.create();
 
-        preparePatterns(recompilationSpec.getClassesToCompile(), classesToDelete, sourceToCompile);
+        prepareJavaPatterns(recompilationSpec.getClassesToCompile(), classesToDelete, sourceToCompile);
         spec.setSourceFiles(narrowDownSourcesToCompile(sourceTree, sourceToCompile));
         includePreviousCompilationOutputOnClasspath(spec);
         addClassesToProcess(spec, recompilationSpec);
         deleteStaleFilesIn(classesToDelete, spec.getDestinationDir());
         deleteStaleFilesIn(classesToDelete, spec.getCompileOptions().getAnnotationProcessorGeneratedSourcesDirectory());
         deleteStaleFilesIn(classesToDelete, spec.getCompileOptions().getHeaderOutputDirectory());
+
+        Map<GeneratedResource.Location, PatternSet> resourcesToDelete = prepareResourcePatterns(recompilationSpec.getResourcesToGenerate(), patternSetFactory);
+        deleteStaleFilesIn(resourcesToDelete.get(GeneratedResource.Location.CLASS_OUTPUT), spec.getDestinationDir());
+        // If the client has not set a location for SOURCE_OUTPUT, javac outputs those files to the CLASS_OUTPUT directory, so clean that instead.
+        deleteStaleFilesIn(resourcesToDelete.get(GeneratedResource.Location.SOURCE_OUTPUT), MoreObjects.firstNonNull(spec.getCompileOptions().getAnnotationProcessorGeneratedSourcesDirectory(), spec.getDestinationDir()));
```

Comment:
This is basically the diciest line in the diff.  We have to cover an implementation detail of `javac`.  See the `SOURCE_OUTPUT` branch in the conditional in [this method](http://hg.openjdk.java.net/jdk/jdk11/file/1ddf9a99e4ad/src/jdk.compiler/share/classes/com/sun/tools/javac/file/JavacFileManager.java#l837).  Doing this might be a bad idea, because it could change over time.

Alternatives include:
1. Not supporting `SOURCE_OUTPUT` (or, any location that is not `CLASS_OUTPUT`) for incremental annotation processing.  I think this is a non-starter because [Filer#createResourceFile](https://docs.oracle.com/en/java/javase/11/docs/api/java.compiler/javax/annotation/processing/Filer.html#createResource(javax.tools.JavaFileManager.Location,java.lang.CharSequence,java.lang.CharSequence,javax.lang.model.element.Element...)) _requires_ that `SOURCE_OUTPUT` be supported.  Of course, not compiling incrementally is not the same as not supporting it, but we can expect that many real processors out in the wild will be using it, and we should make supporting them with incremental compilation a priority.
2. Continuing to support these 3 different `Location`s _but_ disable incremental annotation processing when files with the same relative path are created in more than one of them.  Then, change this code to clean all output directories of all generated resources, instead of trying to match `Location`s to actual directories on disk.  This ought to be an acceptably rare/nonexistent use case in the real world.
The troublesome part is detecting when this happens case so that we can disable incremental compilation: we'd need to collect this data _across_ all of the annotation processors in the compilation task, which is not something we can do currently. For the purposes of incremental compilation, I think we treat each processor in isolation right now.  I'm not sure how hard it'd be to change that.

Reply:
I think doing it like this is the optimal way. It's highly unlikely to ever change in javac.

> The troublesome part is detecting when this happens case so that we can disable incremental compilation: we'd need to collect this data across all of the annotation processors in the compilation task, which is not something we can do currently. For the purposes of incremental compilation, I think we treat each processor in isolation right now. I'm not sure how hard it'd be to change that.

That's not the case, we collect the data for all processors together. Have a look at AnnotationProcessingCompileTask and how it passes the same context object to all processors.

Reply:
Sounds good.  Wasn't sure how bad it is to assume we're wrapping javac.

And you're right, that would be a way to do it.